File: pwaudit/patterns.py
"""Transparent password-weakness pattern detectors.

Each detector looks for one specific, well-understood weakness and, when it
fires, returns a :class:`PatternFinding` whose ``reason`` string explains in
plain language why the pattern hurts the password. These findings are *advisory
downgrade reasons*: they tell the user what to fix. They are deliberately simple
and explainable — unlike the statistical ``zxcvbn`` model, you can read this
file and understand exactly what every check does.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from pwaudit.common_passwords import is_common_password

# --- Keyboard rows used for "keyboard walk" detection ---------------------------
# Each string is one physical row of a QWERTY keyboard, lower-cased.
_KEYBOARD_ROWS: tuple[str, ...] = (
    "1234567890",
    "qwertyuiop",
    "asdfghjkl",
    "zxcvbnm",
)

# Leetspeak substitution map: leet character -> the letter it commonly replaces.
_LEET_MAP: dict[str, str] = {
    "0": "o",
    "1": "i",
    "3": "e",
    "4": "a",
    "5": "s",
    "7": "t",
    "8": "b",
    "9": "g",
    "@": "a",
    "$": "s",
    "!": "i",
    "+": "t",
    "(": "c",
}

# Minimum run length before a sequence / repeat / walk is considered a weakness.
_MIN_RUN = 4


@dataclass(frozen=True)
class PatternFinding:
    """A single detected weakness pattern.

    Attributes:
        code: A short stable identifier (e.g. ``"sequential"``), handy for tests
            and machine-readable output.
        reason: A human-readable explanation of the weakness.
    """

    code: str
    reason: str


def _has_sequential_run(password: str, min_run: int = _MIN_RUN) -> bool:
    """Return ``True`` if *password* contains an ascending or descending run.

    A "run" is ``min_run`` or more consecutive characters whose code points each
    differ from the previous by exactly +1 (ascending) or -1 (descending), e.g.
    ``abcd``, ``4321``, ``wxyz``. The check is case-insensitive.
    """
    pw = password.lower()
    if len(pw) < min_run:
        return False
    asc = desc = 1
    for i in range(1, len(pw)):
        delta = ord(pw[i]) - ord(pw[i - 1])
        asc = asc + 1 if delta == 1 else 1
        desc = desc + 1 if delta == -1 else 1
        if asc >= min_run or desc >= min_run:
            return True
    return False


def _has_repeat_run(password: str, min_run: int = _MIN_RUN) -> bool:
    """Return ``True`` if a single character repeats ``min_run`` times in a row.

    Example: ``aaaa`` or ``1111``. The check is case-insensitive.
    """
    pw = password.lower()
    if len(pw) < min_run:
        return False
    run = 1
    for i in range(1, len(pw)):
        run = run + 1 if pw[i] == pw[i - 1] else 1
        if run >= min_run:
            return True
    return False


def _has_keyboard_walk(password: str, min_run: int = _MIN_RUN) -> bool:
    """Return ``True`` if *password* contains a straight keyboard-row walk.

    A walk is ``min_run`` consecutive characters that appear adjacent, in order
    (forwards or backwards), on a single QWERTY row — e.g. ``qwer``, ``asdf``,
    ``rewq``. The check is case-insensitive.
    """
    pw = password.lower()
    if len(pw) < min_run:
        return False
    for row in _KEYBOARD_ROWS:
        reversed_row = row[::-1]
        for window_start in range(len(pw) - min_run + 1):
            window = pw[window_start : window_start + min_run]
            if window in row or window in reversed_row:
                return True
    return False


def _normalize_leet(password: str) -> str:
    """Replace leetspeak characters in *password* with their plain letters."""
    return "".join(_LEET_MAP.get(c, c) for c in password.lower())


def _is_leetspeak_of_common(password: str) -> str | None:
    """Detect leetspeak obfuscation of a common password.

    If *password* itself contains at least one leet character **and** its
    de-leeted form is a known common password, return that de-leeted word;
    otherwise return ``None``. Requiring an actual substitution avoids
    re-flagging a plain common password (that is reported separately).
    """
    has_leet_char = any(c in _LEET_MAP for c in password.lower())
    if not has_leet_char:
        return None
    normalized = _normalize_leet(password)
    if normalized != password.lower() and is_common_password(normalized):
        return normalized
    return None


# A 4-digit year from 1900 to the near future, used for "year in password".
_YEAR_RE = re.compile(r"(?:19\d\d|20[0-3]\d)")


def _contains_year(password: str) -> str | None:
    """Return the first plausible 4-digit year (1900–2039) found, or ``None``."""
    match = _YEAR_RE.search(password)
    return match.group(0) if match else None


def find_patterns(password: str) -> list[PatternFinding]:
    """Run every pattern detector against *password*.

    Returns a list of :class:`PatternFinding` objects, one per weakness found.
    An empty list means none of the transparent detectors fired (which does not
    by itself mean the password is strong — see :mod:`pwaudit.strength`).
    """
    findings: list[PatternFinding] = []
    if not password:
        return findings

    if is_common_password(password):
        findings.append(
            PatternFinding(
                "common_password",
                "This is one of the most common passwords in breach corpora — "
                "it would be guessed almost immediately.",
            )
        )

    leet_of = _is_leetspeak_of_common(password)
    if leet_of is not None:
        findings.append(
            PatternFinding(
                "leetspeak",
                f"This is leetspeak for the common password '{leet_of}'. "
                "Character swaps like a->@ or o->0 are the first thing crackers try.",
            )
        )

    if _has_sequential_run(password):
        findings.append(
            PatternFinding(
                "sequential",
                "Contains a sequential run (e.g. 'abcd' or '1234'). "
                "Sequences add almost no real randomness.",
            )
        )

    if _has_repeat_run(password):
        findings.append(
            PatternFinding(
                "repeat",
                "Contains a character repeated four or more times in a row "
                "(e.g. 'aaaa'). Repeats barely expand the search space.",
            )
        )

    if _has_keyboard_walk(password):
        findings.append(
            PatternFinding(
                "keyboard_walk",
                "Contains a keyboard walk (e.g. 'qwerty' or 'asdf'). "
                "Adjacent-key patterns are in every cracking wordlist.",
            )
        )

    year = _contains_year(password)
    if year is not None:
        findings.append(
            PatternFinding(
                "year",
                f"Contains the year '{year}'. Years (birthdays, graduations) "
                "are highly predictable and heavily targeted.",
            )
        )

    return findings

File: web/src/patterns.ts
/**
 * Transparent password-weakness pattern detectors (browser side).
 *
 * Each detector targets one well-understood weakness and, when it fires,
 * returns a plain-language reason. These mirror the Python `pwaudit.patterns`
 * detectors so the CLI and the website agree on what counts as a weak pattern.
 * Unlike the statistical zxcvbn model, every check here is small and readable.
 */

/** A single detected weakness pattern. */
export interface PatternFinding {
  /** Stable identifier, e.g. `"keyboard_walk"`. */
  code: string
  /** Human-readable explanation of the weakness. */
  reason: string
}

/** Physical QWERTY keyboard rows, used for keyboard-walk detection. */
const KEYBOARD_ROWS: readonly string[] = [
  '1234567890',
  'qwertyuiop',
  'asdfghjkl',
  'zxcvbnm',
]

/** Leetspeak character -> the letter it commonly replaces. */
const LEET_MAP: Readonly<Record<string, string>> = {
  '0': 'o',
  '1': 'i',
  '3': 'e',
  '4': 'a',
  '5': 's',
  '7': 't',
  '8': 'b',
  '9': 'g',
  '@': 'a',
  $: 's',
  '!': 'i',
  '+': 't',
  '(': 'c',
}

/**
 * A compact set of very common passwords (lower-cased).
 *
 * Deliberately small — a fast "is this obvious" check. The authoritative breach
 * answer comes from the HIBP range API (see {@link file://./hibp.ts}).
 */
const COMMON_PASSWORDS: ReadonlySet<string> = new Set([
  '123456',
  '12345',
  '123456789',
  '12345678',
  '1234567',
  '1234567890',
  '1234',
  '111111',
  '123123',
  '000000',
  '1q2w3e4r',
  '1q2w3e',
  '1qaz2wsx',
  'password',
  'password1',
  'password123',
  'passw0rd',
  'p@ssw0rd',
  'p@ssword',
  'admin',
  'root',
  'letmein',
  'welcome',
  'welcome1',
  'secret',
  'changeme',
  'test',
  'test123',
  'qwerty',
  'qwertyuiop',
  'qwerty123',
  'abc123',
  'abcd1234',
  'iloveyou',
  'monkey',
  'dragon',
  'football',
  'baseball',
  'sunshine',
  'princess',
  'shadow',
  'master',
  'superman',
  'trustno1',
])

const MIN_RUN = 4

/** True if `password` contains an ascending/descending run of `minRun`+. */
function hasSequentialRun(password: string, minRun = MIN_RUN): boolean {
  const pw = password.toLowerCase()
  if (pw.length < minRun) return false
  let asc = 1
  let desc = 1
  for (let i = 1; i < pw.length; i++) {
    const delta = pw.charCodeAt(i) - pw.charCodeAt(i - 1)
    asc = delta === 1 ? asc + 1 : 1
    desc = delta === -1 ? desc + 1 : 1
    if (asc >= minRun || desc >= minRun) return true
  }
  return false
}

/** True if a single character repeats `minRun`+ times in a row. */
function hasRepeatRun(password: string, minRun = MIN_RUN): boolean {
  const pw = password.toLowerCase()
  if (pw.length < minRun) return false
  let run = 1
  for (let i = 1; i < pw.length; i++) {
    run = pw[i] === pw[i - 1] ? run + 1 : 1
    if (run >= minRun) return true
  }
  return false
}

/** True if `password` contains a straight QWERTY-row walk of `minRun`+. */
function hasKeyboardWalk(password: string, minRun = MIN_RUN): boolean {
  const pw = password.toLowerCase()
  if (pw.length < minRun) return false
  for (const row of KEYBOARD_ROWS) {
    const reversed = [...row].reverse().join('')
    for (let start = 0; start <= pw.length - minRun; start++) {
      const window = pw.slice(start, start + minRun)
      if (row.includes(window) || reversed.includes(window)) return true
    }
  }
  return false
}

/** Replace leetspeak characters in `password` with their plain letters. */
function normalizeLeet(password: string): string {
  return [...password.toLowerCase()].map((c) => LEET_MAP[c] ?? c).join('')
}

/**
 * If `password` is leetspeak for a common password, return that common word;
 * otherwise return null. Requires at least one actual leet substitution so a
 * plain common password is not double-reported.
 */
function leetspeakOfCommon(password: string): string | null {
  const hasLeetChar = [...password.toLowerCase()].some((c) => c in LEET_MAP)
  if (!hasLeetChar) return null
  const normalized = normalizeLeet(password)
  if (normalized !== password.toLowerCase() && COMMON_PASSWORDS.has(normalized)) {
    return normalized
  }
  return null
}

/** Plausible 4-digit year, 1900–2039. */
const YEAR_RE = /(?:19\d\d|20[0-3]\d)/

/**
 * Run every pattern detector against `password`.
 *
 * Returns one {@link PatternFinding} per weakness found; an empty array means no
 * transparent detector fired (which does not by itself prove the password is
 * strong — see {@link file://./strength.ts}).
 */
export function findPatterns(password: string): PatternFinding[] {
  const findings: PatternFinding[] = []
  if (!password) return findings

  if (COMMON_PASSWORDS.has(password.toLowerCase())) {
    findings.push({
      code: 'common_password',
      reason:
        'This is one of the most common passwords in breach corpora — it would be guessed almost immediately.',
    })
  }

  const leetOf = leetspeakOfCommon(password)
  if (leetOf !== null) {
    findings.push({
      code: 'leetspeak',
      reason: `This is leetspeak for the common password "${leetOf}". Character swaps like a→@ or o→0 are the first thing crackers try.`,
    })
  }

  if (hasSequentialRun(password)) {
    findings.push({
      code: 'sequential',
      reason:
        'Contains a sequential run (e.g. "abcd" or "1234"). Sequences add almost no real randomness.',
    })
  }

  if (hasRepeatRun(password)) {
    findings.push({
      code: 'repeat',
      reason:
        'Contains a character repeated four or more times in a row (e.g. "aaaa"). Repeats barely expand the search space.',
    })
  }

  if (hasKeyboardWalk(password)) {
    findings.push({
      code: 'keyboard_walk',
      reason:
        'Contains a keyboard walk (e.g. "qwerty" or "asdf"). Adjacent-key patterns are in every cracking wordlist.',
    })
  }

  const yearMatch = password.match(YEAR_RE)
  if (yearMatch) {
    findings.push({
      code: 'year',
      reason: `Contains the year "${yearMatch[0]}". Years (birthdays, graduations) are highly predictable and heavily targeted.`,
    })
  }

  return findings
}
